### PR TITLE
Fix #7 and #9: set up FIMS testing frameworks

### LINKED ISSUE #7: Set up testing system on the FIMS repository
@k-doering-NOAA and I can work on 

- [x] setting up the testing system (e.g., include the files and folders needed) on the FIMS repository
- [x] getting the testing system running on GitHub actions
- [x] adding [C++ files](https://github.com/Bai-Li-NOAA/TestingExamples/blob/main/.gitignore) to the current .gitignore file
- [x] modifying the GH Actions workflow to save artifacts if a job fails

_Originally posted by @Bai-Li-NOAA in https://github.com/NOAA-FIMS/FIMS/issues/3#issuecomment-1049300604_

### PR DESCRIPTION
**Pull request by Bai-Li-NOAA**
Following changes have been made in this pull request:
- set up FIMS testing frameworks by adding tests folder (both R testthat and C++ gtest will be there)
- updated DESCRIPTION and add testthat section
- updated .gitignore file by adding more C++ files and CMake files
- updated .buildignore file to avoid running gtest when doing R CMD check
- added a GH Actions workflow to run google tests

Let me know if you would like me to change anything. Thanks!

### CODE REVIEW

File: .github/workflows/run-googletest.yml
Comment (k-doering-NOAA) [inline .github/workflows/run-googletest.yml:10]: It may be worth adding a CRON trigger, if FIMS has dependencies that we want to make sure are still compatible (e.g., tmb). This could be done later, as needed.

File: .github/workflows/run-googletest.yml
Comment (k-doering-NOAA) [inline .github/workflows/run-googletest.yml:33]: Will the job fail if the Test step have failing tests? We may also want to consider saving artifacts (perhaps conditionally if a job fails), if there is something useful to save (e.g., a detailed summary of which tests failed and why).

File: CMakeLists.txt
Comment (k-doering-NOAA) [inline CMakeLists.txt:13]: would we need to change this from time to time? I wonder if there is a way to just use the latest release? Or maybe fixing to a particular release is intentional?

**Review by k-doering-NOAA — APPROVED** [review]
Thanks for putting this together, @Bai-Li-NOAA! Looks good.

I was also able to run the tests locally by cloning the FIMS repo, checking out the feature branch and running from the command line:

```
cmake -S . -B build -G Ninja
cmake --build build
ctest --test-dir build
```

**Comment by k-doering-NOAA** [comment]
It looks like someone with write access to this repo will also have to approve this PR before it can be merged, btw! (I only have read access).

File: .github/workflows/run-googletest.yml
Comment (Bai-Li-NOAA) [reply .github/workflows/run-googletest.yml:33]: Yes, the job will fail if there are failing tests. Good idea for saving artifacts. I added this suggestion as a new task in issue #7.

File: CMakeLists.txt
Comment (Bai-Li-NOAA) [reply CMakeLists.txt:13]: We will change it, but not very often. We will stick with particular releases when using CMake, Google Test, and Google Benchmark. When there is a need to use newer versions of dependencies, we will conduct regression testing before switching to new versions.

File: inst/include/interface/place_holder0.hpp
Comment (ChristineStawitz-NOAA) [inline inst/include/interface/place_holder0.hpp:45]: Do we need to define this twice (lines 18 - 24 and lines 39 - 45 are identical to my eye)?

File: inst/include/interface/place_holder1.hpp
Comment (ChristineStawitz-NOAA) [inline inst/include/interface/place_holder1.hpp:12]: is it possible to remove this and instead use the `dlognorm` defined in TMB?

File: tests/gtest/test_place_holder2.cpp
Comment (ChristineStawitz-NOAA) [inline tests/gtest/test_place_holder2.cpp:4]: Can we remove the commented out code for now? If we want to include, should do so in a separate file.

**Review by ChristineStawitz-NOAA — APPROVED** [review]
Hi Bai - all of my comments on this are on the cpp side. I just spoke to @Andrea-Havron-NOAA about using a simple logistic function to hook up the `Rcpp` and `TMB` side. I think to avoid duplication we could try removing the logistic growth function here and apply your test case framework to the simpler logistic function @Andrea-Havron-NOAA is using? Or we could merge as-is and let her work on that in a separate pull request. I will approve for merging and leave it up to you.

File: inst/include/interface/place_holder0.hpp
Comment (Bai-Li-NOAA) [reply inst/include/interface/place_holder0.hpp:45]: Good question, I think the lines 18-24 and lines 39-45 are used for different libraries (TMB v.s. std). @Andrea-Havron-NOAA perhaps has a better explanation than me. I copied her FIMS-software [Rcpp code here](https://github.com/Andrea-Havron-NOAA/FIMS-software/blob/master/src/Rcpp/Common.hpp).

File: inst/include/interface/place_holder1.hpp
Comment (Bai-Li-NOAA) [reply inst/include/interface/place_holder1.hpp:12]: I am not sure if TMB has dlognorm function, I know there is a dnorm.hpp. Maybe the name of place_holder is confusing? I plan to remove all testing place_holder.hpp and place_holder.cpp files after the modeling team sets up one example of a module.

File: tests/gtest/test_place_holder2.cpp
Comment (Bai-Li-NOAA) [reply tests/gtest/test_place_holder2.cpp:4]: Those commented out code are for generating one-time true values for C++ unit tests. I agree that for complicated tests we need to put the R script somewhere else, do you have a suggestion on where to put the R script? For simple C++ unit tests, commented out R code is part of the testing documentation. The benefits of including the simple R code as comments in the same C++ testing file is that you can remove or move C++ tests without worrying about where the R code is.

**Comment by Bai-Li-NOAA** [comment]
Hi @ChristineStawitz-NOAA , Totally agree with the logistic function plan. All the testing place_holder files are independent and all of them can be removed once the modeling team sets up one module with some tests. The current testing place_holder files are used for quickly setting up the CMake system and GH actions workflow. @Andrea-Havron-NOAA can set up an independent logistic function without using any of the testing place_holder files. I think it is better to merge the pull request now and let @Andrea-Havron-NOAA work on a new branch (to develop the logistic function) simultaneously with @msupernaw (to modify the CMakeLists.txt).

**Pull request merged**
